File: src/com/android/settings/compatible/CompatibleConfig.java
package com.android.settings.compatible;

import android.content.ContentResolver;
import android.net.Uri;
import android.database.Cursor;
import android.content.ContentValues;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import android.content.Context;

public class CompatibleConfig {
    public static final String COMPATIBLE_URI = "content://com.boringdroid.systemuiprovider";
    public static final String KEY_CODE_IS_ALLOW_SCREENSHOT_AND_RECORD = "isAllowScreenshotAndRecord";
    public static final String KEY_CODE_IS_ALLOW_HIDE_DECOR_CAPTION = "isAllowHideDecorCaption";

    public static Map<String, Object> queryMapValueData(Context context, String packageName, String keycode) {
        Uri uri = Uri.parse(COMPATIBLE_URI + "/COMPATIBLE_VALUE");
        Cursor cursor = null;
        Map<String, Object> result = null;
        String selection = "PACKAGE_NAME = ? AND KEY_CODE = ? AND IS_DEL != 1 ";
        String[] selectionArgs = { packageName, keycode };
        try {
            ContentResolver contentResolver = context.getContentResolver();
            cursor = contentResolver.query(uri, null, selection, selectionArgs, null);
            if (cursor != null && cursor.moveToFirst()) {
                int _ID = cursor.getInt(cursor.getColumnIndex("_ID"));
                String PACKAGE_NAME = cursor.getString(cursor.getColumnIndex("PACKAGE_NAME"));
                String KEY_CODE = cursor.getString(cursor.getColumnIndex("KEY_CODE"));
                String VALUE = cursor.getString(cursor.getColumnIndex("VALUE"));
                String EDIT_DATE = cursor.getString(cursor.getColumnIndex("EDIT_DATE"));
                result = new HashMap<>();
                result.put("_ID", _ID);
                result.put("PACKAGE_NAME", PACKAGE_NAME);
                result.put("KEY_CODE", KEY_CODE);
                result.put("VALUE", VALUE);
                result.put("EDIT_DATE", EDIT_DATE);
            }

        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }
        return result;
    }

    public static Map<String, Object> queryMapValueDataHasDel(Context context, String packageName, String keycode) {
        Uri uri = Uri.parse(COMPATIBLE_URI + "/COMPATIBLE_VALUE");
        Cursor cursor = null;
        Map<String, Object> result = null;
        String selection = "PACKAGE_NAME = ? AND KEY_CODE = ? ";
        String[] selectionArgs = { packageName, keycode };
        try {
            ContentResolver contentResolver = context.getContentResolver();
            cursor = contentResolver.query(uri, null, selection, selectionArgs, null);
            if (cursor != null && cursor.moveToFirst()) {
                int _ID = cursor.getInt(cursor.getColumnIndex("_ID"));
                String PACKAGE_NAME = cursor.getString(cursor.getColumnIndex("PACKAGE_NAME"));
                String KEY_CODE = cursor.getString(cursor.getColumnIndex("KEY_CODE"));
                String VALUE = cursor.getString(cursor.getColumnIndex("VALUE"));
                String EDIT_DATE = cursor.getString(cursor.getColumnIndex("EDIT_DATE"));
                result = new HashMap<>();
                result.put("_ID", _ID);
                result.put("PACKAGE_NAME", PACKAGE_NAME);
                result.put("KEY_CODE", KEY_CODE);
                result.put("VALUE", VALUE);
                result.put("EDIT_DATE", EDIT_DATE);
            }

        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }
        return result;
    }

    public static String queryValueData(Context context, String packageName, String keycode) {
        Uri uri = Uri.parse(COMPATIBLE_URI + "/COMPATIBLE_VALUE");
        Cursor cursor = null;
        String result = null;
        String selection = "PACKAGE_NAME = ? AND KEY_CODE = ?  AND IS_DEL != 1";
        String[] selectionArgs = { packageName, keycode };
        try {
            ContentResolver contentResolver = context.getContentResolver();
            cursor = contentResolver.query(uri, null, selection, selectionArgs, null);
            if (cursor != null && cursor.moveToFirst()) {
                result = cursor.getString(cursor.getColumnIndex("VALUE"));
            }

        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }
        return result;
    }

    public static void insertValueData(Context context, String appName, String packageName, String keycode,
            String value) {
        try {
            Uri uri = Uri.parse(COMPATIBLE_URI + "/COMPATIBLE_VALUE");
            ContentValues values = new ContentValues();
            values.put("PACKAGE_NAME", packageName);
            values.put("KEY_CODE", keycode);
            values.put("VALUE", value);
            values.put("EDIT_DATE", getCurDateTime());
            values.put("FIELDS1", appName);
            values.put("IS_DEL", "0");
            Uri resUri = context.getContentResolver()
                    .insert(uri, values);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static void insertUpdateValueData(Context context, String appName, String packageName, String keycode,
            String value) {
        Map<String, Object> result = queryMapValueDataHasDel(context, packageName, keycode);
        if (result == null) {
            insertValueData(context, appName, packageName, keycode, value);
        } else {
            updateValueData(context, appName, packageName, keycode, value);
        }
    }

    public static int updateValueData(Context context, String appName, String packageName, String keycode,
            String newValue) {
        try {
            Uri uri = Uri.parse(COMPATIBLE_URI + "/COMPATIBLE_VALUE");
            ContentValues values = new ContentValues();
            values.put("VALUE", newValue);
            values.put("FIELDS1", appName);
            values.put("IS_DEL", "0");
            values.put("EDIT_DATE", getCurDateTime());
            String selection = "PACKAGE_NAME = ? AND KEY_CODE = ?";
            String[] selectionArgs = { packageName, keycode };
            int res = context.getContentResolver()
                    .update(uri, values, selection,
                            selectionArgs);
            return res;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return -1;
    }

    public static void deleteValueData(Context context, String packageName, String keycode) {
        try {
            Uri uri = Uri.parse(COMPATIBLE_URI + "/COMPATIBLE_VALUE");
            String selection = "PACKAGE_NAME = ? AND KEY_CODE = ?";
            String[] selectionArgs = { packageName, keycode };
            int res = context.getContentResolver().delete(uri, selection, selectionArgs);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static int deleteValueData(Context context, String packageName) {
        // try {
        // Uri uri = Uri.parse(COMPATIBLE_URI + "/COMPATIBLE_VALUE");
        // String selection = "PACKAGE_NAME = ?";
        // String[] selectionArgs = { packageName };
        // int res = context.getContentResolver().delete(uri, selection, selectionArgs);
        // } catch (Exception e) {
        // e.printStackTrace();
        // }

        try {
            Uri uri = Uri.parse(COMPATIBLE_URI + "/COMPATIBLE_VALUE");
            ContentValues values = new ContentValues();
            values.put("IS_DEL", "1");
            values.put("EDIT_DATE", getCurDateTime());
            String selection = "PACKAGE_NAME = ? ";
            String[] selectionArgs = { packageName };
            int res = context.getContentResolver()
                    .update(uri, values, selection,
                            selectionArgs);
            return res;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return -1;
    }

    public static void cleanValueData(Context context) {
        try {
            Uri uri = Uri.parse(COMPATIBLE_URI + "/COMPATIBLE_VALUE");
            String selection = null;
            String[] selectionArgs = null;
            int res = context.getContentResolver().delete(uri, selection, selectionArgs);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static List<Map<String, Object>> queryValueListData(Context context, String keycode) {
        Uri uri = Uri.parse(COMPATIBLE_URI + "/COMPATIBLE_VALUE");
        List<Map<String, Object>> list = null;
        Cursor cursor = null;
        String selection = " KEY_CODE = ?  AND IS_DEL != 1";
        String[] selectionArgs = { keycode };
        try {

            ContentResolver contentResolver = context.getContentResolver();
            cursor = contentResolver.query(uri, null, selection, selectionArgs, null);
            if (cursor != null && cursor.moveToFirst()) {
                list = new ArrayList<>();
                do {
                    int _ID = cursor.getInt(cursor.getColumnIndex("_ID"));
                    String PACKAGE_NAME = cursor.getString(cursor.getColumnIndex("PACKAGE_NAME"));
                    String KEY_CODE = cursor.getString(cursor.getColumnIndex("KEY_CODE"));
                    String EDIT_DATE = cursor.getString(cursor.getColumnIndex("EDIT_DATE"));
                    String VALUE = cursor.getString(cursor.getColumnIndex("VALUE"));
                    String NOTES = cursor.getString(cursor.getColumnIndex("NOTES"));
                    String FIELDS1 = cursor.getString(cursor.getColumnIndex("FIELDS1"));
                    Map<String, Object> mp = new HashMap<>();
                    mp.put("_ID", _ID);
                    mp.put("PACKAGE_NAME", PACKAGE_NAME);
                    mp.put("VALUE", VALUE);
                    mp.put("KEY_CODE", KEY_CODE);
                    mp.put("NOTES", NOTES);
                    mp.put("FIELDS1", FIELDS1);
                    mp.put("EDIT_DATE", EDIT_DATE);
                    list.add(mp);
                } while (cursor.moveToNext());
            }

        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }
        return list;
    }

    public static List<Map<String, Object>> queryListData(Context context) {
        Uri uri = Uri.parse(COMPATIBLE_URI + "/COMPATIBLE_LIST");
        List<Map<String, Object>> list = null;
        Cursor cursor = null;
        String selection = null;
        String[] selectionArgs = null;
        try {

            ContentResolver contentResolver = context.getContentResolver();
            cursor = contentResolver.query(uri, null, selection, selectionArgs, null);
            if (cursor != null && cursor.moveToFirst()) {
                list = new ArrayList<>();
                do {
                    int _ID = cursor.getInt(cursor.getColumnIndex("_ID"));
                    String KEY_CODE = cursor.getString(cursor.getColumnIndex("KEY_CODE"));
                    String KEY_DESC = cursor.getString(cursor.getColumnIndex("KEY_DESC"));
                    String CREATE_DATE = cursor.getString(cursor.getColumnIndex("CREATE_DATE"));
                    String DEFAULT_VALUE = cursor.getString(cursor.getColumnIndex("DEFAULT_VALUE"));
                    String OPTION_JSON = cursor.getString(cursor.getColumnIndex("OPTION_JSON"));
                    String INPUT_TYPE = cursor.getString(cursor.getColumnIndex("INPUT_TYPE"));
                    String NOTES = cursor.getString(cursor.getColumnIndex("NOTES"));
                    Map<String, Object> mp = new HashMap<>();
                    mp.put("_ID", _ID);
                    mp.put("DEFAULT_VALUE", DEFAULT_VALUE);
                    mp.put("OPTION_JSON", OPTION_JSON);
                    mp.put("KEY_CODE", KEY_CODE);
                    mp.put("KEY_DESC", KEY_DESC);
                    mp.put("NOTES", NOTES);
                    mp.put("CREATE_DATE", CREATE_DATE);
                    mp.put("INPUT_TYPE", INPUT_TYPE);
                    list.add(mp);
                } while (cursor.moveToNext());
            }

        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }
        return list;
    }

    public static void recoveryValueData(Context context, String packageName, String keycode) {
        try {
            Uri uri = Uri.parse(COMPATIBLE_URI + "/RECOVERY_VALUE");
            ContentValues values = new ContentValues();
            values.put("PACKAGE_NAME", packageName);
            values.put("KEY_CODE", keycode);
            values.put("IS_DEL", "0");
            Uri resUri = context.getContentResolver()
                    .insert(uri, values);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static String getCurDateTime() {
        LocalDateTime currentTime = LocalDateTime.now();
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
        String formattedTime = currentTime.format(formatter);
        return formattedTime;
    }

}

File: src/com/android/settings/compatible/CompatibleConfig.kt
package com.android.settings.compatible;

import android.content.Context
import android.database.Cursor
import android.net.Uri
import android.content.ContentValues
import com.android.settings.R;

object CompatibleConfig {
    const val COMPATIBLE_STR = "com.android.compatibleprovider"
    const val COMPATIBLE_URI = "content://$COMPATIBLE_STR"
    const val TAG = "CompatibleConfig"


    fun queryMapValueData(
        context: Context,
        selection: String,
        selectionArgs: Array<String>
    ): Map<String?, Any?>? {
        val uri = Uri.parse("$COMPATIBLE_URI/COMPATIBLE_VALUE")
        var cursor: Cursor? = null
        var result: MutableMap<String?, Any?>? = null
        val queryParam = "$selection AND IS_DEL != 1"


        try {
            val contentResolver = context.contentResolver
            cursor = contentResolver.query(uri, null, queryParam, selectionArgs, null)
            if (cursor != null && cursor.moveToFirst()) {
                val packageName = cursor.getString(cursor.getColumnIndexOrThrow("PACKAGE_NAME"))
            }
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
        } finally {
            cursor?.close()
        }
        return result
    }

    fun queryValueListByKeyCode(
        context: Context,
        keyCode: String
    ): List<CompatibleValue>? {
        val uri = Uri.parse(COMPATIBLE_URI + "/COMPATIBLE_VALUE")
        var cursor: Cursor? = null
        val selection = "KEY_CODE = ? AND IS_DEL != 1"
        val selectionArgs = arrayOf(keyCode)
        try {
            val contentResolver = context.contentResolver
            cursor = contentResolver.query(uri, null, selection, selectionArgs, null)
            var list = ArrayList<CompatibleValue>()
            if (cursor != null && cursor.moveToFirst()) {
                do {
                    val id = cursor.getInt(cursor.getColumnIndexOrThrow("_ID"))
                    val packageName = cursor.getString(cursor.getColumnIndexOrThrow("PACKAGE_NAME"))
                    val keyCode = cursor.getString(cursor.getColumnIndexOrThrow("KEY_CODE"))
                    val value = cursor.getString(cursor.getColumnIndexOrThrow("VALUE"))
                    // val appName = cursor.getString(cursor.getColumnIndexOrThrow("APP_NAME"))
                    val activityName =
                        cursor.getString(cursor.getColumnIndexOrThrow("ACTIVITY_NAME"))
                    val createDate = cursor.getString(cursor.getColumnIndexOrThrow("CREATE_DATE"))
                    val editDate = cursor.getString(cursor.getColumnIndexOrThrow("EDIT_DATE"))
                    val fields1 = cursor.getString(cursor.getColumnIndexOrThrow("FIELDS1"))

                    var compatibleValue: CompatibleValue = CompatibleValue();
                    compatibleValue.id = id
                    compatibleValue.packageName = packageName
                    compatibleValue.keyCode = keyCode
                    compatibleValue.value = value
//                    compatibleValue.appName = appName
                    compatibleValue.activityName = activityName
                    compatibleValue.createDate = createDate
                    compatibleValue.editDate = editDate
                    compatibleValue.fields1 = fields1

                    val appData = CompUtils.getAppInfo(context, packageName)
                    if (appData != null) {
                        compatibleValue.appName = appData.name.toString();
                        list.add(compatibleValue)
                    }
                } while (cursor.moveToNext())
            }
            return list;
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            cursor?.close()
        }
        return null
    }


    fun queryValueListByKeyCodeAndPackageName(
        context: Context,
        packageName: String,
        keyCode: String
    ): MutableList<CompatibleValue>? {
        val uri = Uri.parse(COMPATIBLE_URI + "/COMPATIBLE_VALUE")
        var cursor: Cursor? = null
        val selection = "PACKAGE_NAME = ? AND KEY_CODE = ? AND IS_DEL != 1"
        val selectionArgs = arrayOf(packageName, keyCode)
        try {
            val contentResolver = context.contentResolver
            cursor = contentResolver.query(uri, null, selection, selectionArgs, null)
            var list : MutableList<CompatibleValue> = mutableListOf() 
            if (cursor != null && cursor.moveToFirst()) {
                do {
                    val id = cursor.getInt(cursor.getColumnIndexOrThrow("_ID"))
                    val packageName = cursor.getString(cursor.getColumnIndexOrThrow("PACKAGE_NAME"))
                    val keyCode = cursor.getString(cursor.getColumnIndexOrThrow("KEY_CODE"))
                    val value = cursor.getString(cursor.getColumnIndexOrThrow("VALUE"))
                    // val appName = cursor.getString(cursor.getColumnIndexOrThrow("APP_NAME"))
                    val activityName =
                        cursor.getString(cursor.getColumnIndexOrThrow("ACTIVITY_NAME"))
                    val createDate = cursor.getString(cursor.getColumnIndexOrThrow("CREATE_DATE"))
                    val editDate = cursor.getString(cursor.getColumnIndexOrThrow("EDIT_DATE"))
                    val fields1 = cursor.getString(cursor.getColumnIndexOrThrow("FIELDS1"))

                    var compatibleValue: CompatibleValue = CompatibleValue();
                    compatibleValue.id = id
                    compatibleValue.packageName = packageName
                    compatibleValue.keyCode = keyCode
                    compatibleValue.value = value
                    // compatibleValue.appName = appName
                    compatibleValue.activityName = activityName
                    compatibleValue.createDate = createDate
                    compatibleValue.editDate = editDate
                    compatibleValue.fields1 = fields1
                    list.add(compatibleValue)
                } while (cursor.moveToNext())
            }
            return list;
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            cursor?.close()
        }
        return null
    }


    fun queryMapValueData(
        context: Context,
        keyCode: String,
        packageName: String,
        acitivityName: String
    ): CompatibleValue? {
        val uri = Uri.parse(COMPATIBLE_URI + "/COMPATIBLE_VALUE")
        var cursor: Cursor? = null
        val selection = "PACKAGE_NAME = ? AND KEY_CODE = ? AND ACTIVITY_NAME = ? AND IS_DEL != 1"
        val selectionArgs = arrayOf(packageName, keyCode, acitivityName)
        try {
            val contentResolver = context.contentResolver
            cursor = contentResolver.query(uri, null, selection, selectionArgs, null)
            if (cursor != null && cursor.moveToFirst()) {
                val id = cursor.getInt(cursor.getColumnIndexOrThrow("_ID"))
                val packageName = cursor.getString(cursor.getColumnIndexOrThrow("PACKAGE_NAME"))
                val keyCode = cursor.getString(cursor.getColumnIndexOrThrow("KEY_CODE"))
                val value = cursor.getString(cursor.getColumnIndexOrThrow("VALUE"))
                val appName = cursor.getString(cursor.getColumnIndexOrThrow("APP_NAME"))
                val activityName =
                    cursor.getString(cursor.getColumnIndexOrThrow("ACTIVITY_NAME"))
                val createDate = cursor.getString(cursor.getColumnIndexOrThrow("CREATE_DATE"))
                val editDate = cursor.getString(cursor.getColumnIndexOrThrow("EDIT_DATE"))
                val fields1 = cursor.getString(cursor.getColumnIndexOrThrow("FIELDS1"))

                var compatibleValue: CompatibleValue = CompatibleValue();
                compatibleValue.id = id
                compatibleValue.packageName = packageName
                compatibleValue.keyCode = keyCode
                compatibleValue.value = value
                compatibleValue.appName = appName
                compatibleValue.activityName = activityName
                compatibleValue.createDate = createDate
                compatibleValue.editDate = editDate
                compatibleValue.fields1 = fields1
                return  compatibleValue ;
            }
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            cursor?.close()
        }
        return null
    }

    fun insertValueData(
        context: Context, keyCode: String, packageName: String, activityName: String,
        value: String
    ): Int {
        try {
            var newValue = value;
            if(newValue.equals(context.getString(R.string.fde_compatible_unset))){
                newValue = "";
            }
            if (newValue.contains("x")) {
                 newValue = CompUtils.sizeStringToJson(newValue);   
            }
            if (activityName == null || "".equals(activityName)) {
                CompUtils.setSystemProperty(packageName + "_" + keyCode, newValue)
            } else {
                CompUtils.setSystemProperty(
                    packageName + "_" + keyCode + "_" + activityName,
                    newValue
                )
            }
            val uri = Uri.parse(COMPATIBLE_URI + "/COMPATIBLE_VALUE")
            val values = ContentValues()
            val curTime = CompUtils.getCurDateTime();
            values.put("PACKAGE_NAME", packageName)
            values.put("KEY_CODE", keyCode)
            values.put("IS_ENABLE", "1");
            values.put("APP_NAME", "");
            values.put("VALUE", newValue)
            values.put("IS_DEL", "0")
            values.put("CREATE_DATE", curTime);
            values.put("EDIT_DATE", curTime);
            values.put("FIELDS1", "");
            values.put("FIELDS2", "");
            values.put("ACTIVITY_NAME", activityName);
            values.put("EDIT_DATE", curTime)
            val selection = "PACKAGE_NAME = ? AND KEY_CODE = ? AND ACTIVITY_NAME = ?"
            val selectionArgs = arrayOf(packageName, keyCode, activityName)
            context.contentResolver
                .insert(
                    uri, values
                )
            return 0
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
        }
        return -1
    }


    fun updateValueData(
        context: Context, keyCode: String, packageName: String, acitivityName: String,
        value: String
    ): Int {
        try {
            var newValue = value;
            if(newValue.equals(context.getString(R.string.fde_compatible_unset))){
                newValue = "";
            }
            if (newValue.contains("x")) {
                 newValue = CompUtils.sizeStringToJson(newValue);   
            }
            if (acitivityName == null || "".equals(acitivityName)) {
                CompUtils.setSystemProperty(packageName + "_" + keyCode, newValue)
            } else {
                CompUtils.setSystemProperty(
                    packageName + "_" + keyCode + "_" + acitivityName,
                    newValue
                )
            }
            val uri = Uri.parse(COMPATIBLE_URI + "/COMPATIBLE_VALUE")
            val values = ContentValues()
            values.put("VALUE", newValue)
            values.put("IS_DEL", "0")
            values.put("EDIT_DATE", CompUtils.getCurDateTime())
            val selection = "PACKAGE_NAME = ? AND KEY_CODE = ? AND ACTIVITY_NAME = ?"
            val selectionArgs = arrayOf(packageName, keyCode, acitivityName)
            val res = context.contentResolver
                .update(
                    uri, values, selection,
                    selectionArgs
                )
            return res
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
        }
        return -1
    }

}